### ISSUE TITLE
App crash when opening any files in any git repository using index version 4

### ISSUE THREAD

**Issue by niw** [atom/atom#11225]
### Description

Due to an issue in `libgit2` used from `git-utils` in Atom editor dependency, when opening any files in any git repository using index version 4 may crash the app.

See details at atom/git-utils#70 as well.

The crash may be mitigated by atom/git-utils#71 and pick that change to Atom editor, however, it doesn't mean `git-utils` supports git repository using index version 4. Need to support index version 4 by `libgit2`.

### Steps to Reproduce

1. Create a git repository and commit a file.
2. Run git update-index --index-version=4 to update index version to 4.
3. Launch Atom and open any file in that repository.

**Expected behavior:**
Just open that file.

**Actual behavior:**
Crash

### Versions
    $ atom --version
    1.6.0

    $ apm --version
    apm  1.6.0
    npm  2.13.3
    node 0.10.40
    python 2.7.10
    git 2.7.4

**Comment by lee-dohm** [comment]
It seems that this is a duplicate of atom/git-utils#70. Am I mistaken?

**Comment by niw** [comment]
@lee-dohm It's slightly different, as I tested, atom/git-utils#70 may be solved by atom/git-utils#71 but Atom editor may still crash with the other issues, so I suggest to trace this issue separately.

I think Atom (and also `git-utils`) may properly need to handle the case which `libgit2` can't read index file.

**Comment by lee-dohm** [comment]
Can you update the title to reflect the different context? I'm not sure what the best name for it would be 😕

**Comment by niw** [comment]
@lee-dohm I think the title is correct and describes the exact issue of Atom editor. My expectation in descriptions that atom/git-utils#71 may fix this crash as well was not correct so I've updated it.
Also, atom/git-utils#70 title is correct, the difference is "(Atom editor) crashes by opening files in the such git repository" vs. "(Node process) crashes by opening such git repository in git-utils API."

**Comment by xcv58** [comment]
I build Atom with local fork git-utils which update libgit2 to 0.24.1

It works for me. But we may need update of git-utils.

**Comment by niw** [comment]
@xcv58 yes, at least updating `libgit2` to 0.24.0 (or 0.24.1) prevents an instant atom crash with git repository using index version 4.

libgit2/libgit2#3837 is the real solution for this issue and it's on `libgit2` master.
As I mentioned on atom/git-utils#70, probably it might be better to wait next `libgit2` release, then update atom/git-utils, then update atom/atom.

**Comment by niw** [comment]
atom/git-utils#70 is merged also atom/git-utils@3b11d25 bumped libgit2/libgit2 version to 0.24.3, atom/git-utils v4.1.3 prevents atom/atom crashes when opening git repository using v4 index.
However, it's still not working properly and we need to wait libgit2/libgit2 version to 0.25.0 which will have a real solution libgit2/libgit2#3837.
